src/stirling/source_connectors/socket_tracer/socket_trace_connector.cc

```diff
@@ -845,6 +904,158 @@ void SocketTraceConnector::AcceptHTTP2Data(std::unique_ptr<HTTP2DataEvent> event
   tracker.AddHTTP2Data(std::move(event));
 }
 
+void SocketTraceConnector::AcceptGrpcCEvent(struct conn_id_t connection_id, uint32_t stream_id,
+                                            uint64_t timestamp, bool outgoing,
+                                            uint64_t position_in_stream,
+                                            std::vector<struct grpc_c_data_slice_t> slices) {
+  // Initiate data event template as if it arrived from Golang gRPC eBPF.
+  struct go_grpc_data_event_t data_event_go_style = {};
+
+  data_event_go_style.attr.end_stream = false;  // End stream will be on an empty header, not data.
+  data_event_go_style.attr.stream_id = stream_id;
+  if (outgoing == true) {
+    data_event_go_style.attr.event_type = kDataFrameEventWrite;
+  } else {
+    data_event_go_style.attr.event_type = kDataFrameEventRead;
+  }
+  data_event_go_style.data_attr.pos = position_in_stream;
+  data_event_go_style.attr.timestamp_ns = timestamp;
+  data_event_go_style.attr.probe_type = k_probe_type_other;
+  data_event_go_style.attr.conn_id = connection_id;
+
+  uint64_t extra_position = 0;
+  for (struct grpc_c_data_slice_t data_slice : slices) {
+    struct go_grpc_data_event_t current_data_event = data_event_go_style;
+    current_data_event.data_attr.pos = position_in_stream += extra_position;
+    extra_position += data_slice.length;  // Add extra position for next slices.
+    current_data_event.data_attr.data_size = data_slice.length;
+    current_data_event.data_attr.data_buf_size =
+        std::min((uint32_t)data_slice.length, (uint32_t)GRPC_C_SLICE_SIZE);
+    if ((uint32_t)data_slice.length > (uint32_t)GRPC_C_SLICE_SIZE) {
+      LOG(WARNING) << absl::Substitute("gRPC-c $0 slice of size $1 is larger than maximum size $2",
+                                       outgoing ? "outgoing" : "incoming",
+                                       (uint32_t)data_slice.length, (uint32_t)GRPC_C_SLICE_SIZE);
+    }
+
+    auto d_event_ready = std::make_unique<HTTP2DataEvent>();
+    try {
+      d_event_ready->attr = current_data_event.attr;
+      d_event_ready->data_attr = current_data_event.data_attr;
+      // Assign instead of copy so null terminator is copied too.
+      // Assign can throw when wrong length is given, and since the data is brought from eBPF,
+      // better catch.
+      d_event_ready->payload =
+          std::string_view(data_slice.bytes, current_data_event.data_attr.data_buf_size);
+    } catch (const std::exception& e) {
+      LOG(WARNING) << absl::Substitute("gRPC-C slice data assigning threw exception $0", e.what());
+      return;
+    }
```

Comment:
Pixie C++ code don't use exception, following the google's style guide.
Where exception can be thrown? It seems everything is just assignment of PoD types, which should not throw.

Reply:
Done